### DISCUSSION TITLE
Notifications / webhooks ?

### DISCUSSION THREAD

**Discussion by lukasMega** [FreshRSS/FreshRSS#6480] [Q&A]
Hello, 
I'm still quite new in FreshRSS and I apologize in advance if there is already solution for my question. I have searched, but I have not found anything.

Q: Is possible to integrate some notification system (email, ntfy, gotify, ...) or make webhook when some specific article arises (based on my custom query/filters) ?

I would like to simply call a HTTP endpoint with article info so other service will receive it and it can process later (for example send notification).

<details><summary><b>Try to explain in more details:</b></summary>

```
              ┌──────────────────┐                                ┌────────────────┐                        
              │                  │                                │                │                        
              │                  │                                │     Gotify     │                        
              │     FreshRSS     │                                │                │                        
              │                  │                                │ or another     │                        
              │                  │ checks RSS periodically        │        service │                        
              └────────┬─────────┘ for new articles               └───────┬────────┘                        
                       │                     │                            │                                 
                       │                     │                            │                                 
                       │                     │                            │                                 
                       ├────────────────────►│                            │                                 
                       │                     │                            │                                 
                       │◄────────────────────┤                            │                                 
                       │                     │                            │                                 
                       │                     │                            │                                 
                       │                     │                            │                                 
                       │                     │                            │                                 
                       │                     │                            │                                 
                       │                     │                            │                                 
                       │                     │                            │                                 
                       │                     │                            │                                 
                       │                     │                            │                                 
                       ├────────────────────►│                            │                                 
                       │                     │                            │                                 
                       │◄────────────────────┤                            │                                 
when some new article  │                     │                            │                                 
(that matches          │                                                  │                                 
custom filters) appears├────────────────────────────────────────────────► │                                 
                       │              webhook     https://gotify.github.io/api-docs/#/message/createMessage 
                       │                                                  │                                 
                       │                                                  │                                 
                       │                                                  │                                 
                       ▼                                                  ▼                                 
                                                                                                            
```
</details>

- FreshRSS will send HTTP request to user(admin)-defined URL. It must be configurable:
  - *HTTP Method: GET, POST, PUT, …
  - *URL
  - body (JSON should be enough)
  - optional: Headers

\* - required options

---

Thank you

**Comment by Alkarex** [comment]
Hello,
At the moment, there is no such notification mechanism built in. However, you can [produce an RSS feed with any search of interest](https://freshrss.github.io/FreshRSS/en/users/user_queries.html#share-your-user-queries), and point to that RSS feed from a third party alerting tool, for instance https://ifttt.com/connect/feed/if_notifications

**Reply by lukasMega** [reply]
Hmm, I found this workaround. But I think it will be great and more efficient to not involve some other 3rd service/software for it. 
Because such service as ifttt **needs to check periodically** FreshRSS endpoint and then it calls desired service (as I mentioned for example Gotify, ntfy, email, …)
 - instead, FreshRSS can make request itself only when needed (when all conditions are met)

That's my opinion.

I will try to create a simple plugin for it. I found one existing plugin for inspiration which uses `entry_before_insert` hook.
 - but I still think that should be the core functionality in FreshRSS

**Comment by lukasMega** [comment]
I created an extension for myself.
And it's not finished yet, but it's working (:

There are quite a lot of settings:


<details><summary> 👉 click to see screenshot</summary>

<img width="639" alt="Screen 2024-05-24 at 17 14 10" src="https://github.com/FreshRSS/FreshRSS/assets/2448004/123fe48f-d5fd-4639-abc6-4c9d73ba5873">
</details>

<details><summary> 👉 click to see another similar screenshot in other theme</summary>

<img width="763" alt="Screen 2024-05-24 at 17 56 07" src="https://github.com/FreshRSS/FreshRSS/assets/2448004/5b190264-14ee-41ba-ab42-7d310a00c159">
</details>

<details><summary> 👉 click to see  how it looks when Gotify received webhook:</summary>

<img width="438" alt="Screen 2024-05-24 at 17 51 55" src="https://github.com/FreshRSS/FreshRSS/assets/2448004/10493b12-d8c1-4fb0-b58e-96f24c9b3173">
</details>



<details><summary> 👉 click to see  how it works:</summary>

## How it works

```
┌──────────────┐          ┌────────────────────────────────────┐          ┌───────┐
│              │          │             FreshRSS               │          │       │
│              │          │                                    │          │ some  │
│   INTERNET   │          │  ┌────────┐         ┌───────────┐  │          │       │
│              │          │  │FreshRSS│         │  Webhook  │  │          │service│
│              │          │  │  core  │         │ extension │  │          │       │
└────┬─────────┘          └──┴──┬─────┴─────────┴─────┬─────┴──┘          └─────┬─┘
     │                          │                     │                         │  
     │       checks RSS         │                     │                         │  
     │     for new articles     │                     │                         │  
     │◄─────────────────────────┤                     │                         │  
     │                          │                     │ if some new article     │  
     ├─────────────────────────►│                     │ matches custom criteria │  
     │       new articles       ├────────────────────►│                         │  
     │                          │     new articles    ├────────────────────────►│  
     │                          │                     │      HTTP request       │  
     │                          │                     │                         │  
     │      checks RSS          │                     │                         │  
     │     or new articles      │                     │                         │  
     │◄─────────────────────────┤                     │                         │  
     │                          │                     │ if no new article       │  
     ├─────────────────────────►│                     │ matches custom criteria │  
     │      new articles        ├────────────────────►│ no request will be sent │  
     │                          │     new articles    │                         │  
     │                          │                     │                         │  
     │                          │                     │                         │  
     ▼                          ▼                     ▼                         ▼  
```

- for every new article that matches custom criteria new HTTP request will be sent

## ⚠️ Limitations

- currently only GET, POST and PUT methods are supported
- there are no validation for configuration
- only one webhook, but it's enough for me.

</details>

**Reply by Alkarex** [reply]
It looks very good already 🙂
Feel free to share a PR on
https://github.com/FreshRSS/Extensions

**Reply by lukasMega** [reply]
Small update. Probably next week I will do a PR.

It's working at least for 90%. I need finish some details and edge cases.

Right now I'm using `Minz_Log::warning` for informational logging purposes in order to display logs in production environment. I plan to remove such logs.

Testing webhooks are done in Node-RED (https://github.com/node-red/node-red-docker):

 -  ![Webhook_FreshRSS_0 0 1](https://github.com/FreshRSS/FreshRSS/assets/2448004/82f92588-68b7-4d95-b259-e3f738e3c5b7)

**Reply by adrianmusante** [reply]
Any news on this? It looks incredible. I saw a PR you made but nothing to do with this.

**Reply by lukasMega** [reply]
> Any news on this? It looks incredible. I saw a PR you made but nothing to do with this.

It seems, there is some demand for such webhook functionality. I didn't made a PR yet. But I will try to create a one. I didn't have a time to test and ensure the correctness for all the possibilities, but it's already working a few months for my needs.

PHP is not my favourite language, actually I hate PHP, but I love FreshRSS and I convinced myself to implement this.
I can create a PR in next week, but I would be grateful to someone experienced with PHP to finish it (make the code better, I mean to align the code to follow php conventions).
Thank you for understanding.

**Reply by cmuench** [reply]
> > Any news on this? It looks incredible. I saw a PR you made but nothing to do with this.
> 
> It seems, there is some demand for such webhook functionality. I didn't made a PR yet. But I will try to create a one. I didn't have a time to test and ensure the correctness for all the possibilities, but it's already working a few months for my needs.
> 
> PHP is not my favourite language, actually I hate PHP, but I love FreshRSS and I convinced myself to implement this. I can create a PR in next week, but I would be grateful to someone experienced with PHP to finish it (make the code better, I mean to align the code to follow php conventions). Thank you for understanding.

If you think it's not 100% ready, you can create a draft PR. I tink there will be some guys who can review the code.

**Reply by lukasMega** [reply]
I will create a draft PR soon :) 
... update: https://github.com/FreshRSS/Extensions/pull/249
please ensure you have enough patience, since I don't have a PHP environment locally and things like PHPStan I saw first time when I checked github action for my PR (:

**Reply by lukasMega** [reply]
I apologize for my long inactivity, I have no time due to family obligations.

I would be very glad if someone could take over (https://github.com/FreshRSS/Extensions/pull/249) and enhance the code to address any issues and ensure it meets the necessary criteria for inclusion in this repository. ❤️

**Reply by Ryahn** [reply]
> I apologize for my long inactivity, I have no time due to family obligations.
> 
> I would be very glad if someone could take over ([FreshRSS/Extensions#249](https://github.com/FreshRSS/Extensions/pull/249)) and enhance the code to address any issues and ensure it meets the necessary criteria for inclusion in this repository. ❤️

I have make a new PR with https://github.com/FreshRSS/Extensions/pull/324
